relay-server/src/actors/processor.rs

```
@@ -2279,59 +2284,92 @@ impl EnvelopeProcessorService {
-            let validated_span = match self.validate_span(annotated_span) {
-                Ok(span) => span,
+        let received_at = state.managed_envelope.received_at();
+
+        state.managed_envelope.retain_items(|item| {
+            let annotated_span = match item.ty() {
+                ItemType::OtelSpan => {
+                    match serde_json::from_slice::<relay_spans::OtelSpan>(&item.payload()) {
+                        Ok(otel_span) => Annotated::new(otel_span.into()),
+                        Err(err) => {
+                            relay_log::debug!("failed to parse OTel span: {}", err);
+                            return ItemAction::DropSilently;
+                        }
+                    }
+                }
+                ItemType::Span => match Annotated::<Span>::from_json_bytes(&item.payload()) {
+                    Ok(span) => span,
+                    Err(err) => {
+                        relay_log::debug!("failed to parse span: {}", err);
+                        return ItemAction::DropSilently;
+                    }
+                },
+
+                _ => return ItemAction::Keep,
+            };
+
+            let config = NormalizeSpanConfig {
+                received_at,
+                max_secs_in_past: self.inner.config.max_secs_in_past(),
+                max_secs_in_future: self.inner.config.max_secs_in_future(),
+                max_tag_value_length: self
+                    .inner
+                    .config
+                    .aggregator_config_for(MetricNamespace::Spans)
+                    .max_tag_value_length,
+            };
+            let mut annotated_span = match normalize_span(annotated_span, config) {
+                Ok(s) => s,
                 Err(e) => {
-                    relay_log::error!("invalid span: {e}");
-                    return;
+                    relay_log::debug!("failed to normalize span: {}", e);
+                    return ItemAction::DropSilently;
```

Why the error -> debug downgrade, considering that spans are silently dropped?

---

Good question! `normalize_span` returns an error when the span is invalid (e.g. missing trace ID), so we don't want to be notified about that. I changed the wording to make the failure case more clear.